rfcs/XXXX-gene-sets/text.md

```diff
@@ -0,0 +1,97 @@
+are selected and the difference in the expression level for each gene is computed by comparing the average
+expression level for each set of cells. When there is a particularly large difference in expression between the two sets of
+cells, that gene is identified and presented to the user. To ensure consistency and compatibility it is necessary to define a
+data format for storing references to these genes and any accompanying statistical data.
+
+Product Requirements
+
+### Users should be able to export differential expression results
+
+Currently cellxgene users can choose two sets of cells and compute/display the log fold and adjusted p-value of the top ten differentially expressed genes. Additionally we display a histogram of the gene expression levels of the selected cells for each gene in the right side bar.
+
+- [US1] A user should be able to export a gene set, if the gene set was created by differential expression, the pvalue, log fold change, and category label fields will be populated in addition to the gene name. This will be available in a csv file (see data formats below for more detail).
+
+### Users should be able to upload their own gene sets and see them in the right side bar with the option to apply them within current cellxgene functionalities, if the same gene ontology was used when generating the current dataset and the geneset uploaded by the user (ie the gene names match)
+
+- [US2] A user should be able to upload a geneset via a csv file (see data formats below for more detail) for a dataset.
+- [US3] A user should be able to upload a geneset they previously downloaded from cellxgene
+- [US4] When a user uploads a geneset, that list of genes should appear in the right side bar
+- [US5] If the user was logged in when they uploaded a geneset that genesit should persist if they logout and back in or reload the page
+- [US6] If the gene names match those used in the dataset the user should be able to use general cellxgene functionality (color by genes, calculate differential expression levels, etc.) on those genes
+
+## Detailed Design | Architecture | Implementation
+
+### Data model
+
+Fully supporting persistence of differential expression results is out of the scope of this RFC. But the data model described below has been designed to allow for differential expression persistence in the future.
+
+### Relational Database
+
+To support gene sets three tables will be added to the cellxgene relational database
+
+- Gene
+  - UUID (generated)
+  - GeneName
+  - OntologyId?
+- GeneGeneSetLink
+  - UUID
+  - Gene UUID
+  - GeneSet UUID
+  - Comments (optional)
+- GeneSet
+  - UUID (generated)
+  - GeneSet Name (required)
+  - Gene Count (generated)
+  - User UUID
+  - Dataset UUID
+
+![Cellxgene Data Schema](imgs/Cellxgene_rds_schema.png)
+**Figure 1** Cellxgene data schema, tables to be added (as described above) are in red
+
+#### Imported CSV
+
+GeneSet CSV
```

Comment:
Few questions.

1.

Where would the gene comments come from?   Will we have a way to associate a gene with a comment for export, or is this only an import format issue.   There isn't a place for gene comments in the db schema.

2.

The CELLSET_CATEGORY.LABEL scheme has some limitations.  A common case is a user wants to run differential expressions with a set of cell, and all the rest.   A vs NOT A.  How could we encode that?

Furthermore, as a generalization you could imagine some utility in doing unions of labels.  union(A,B) vs union(C,D).

3.

What parts of this file are read into cellxgene on import?  Just the  GENESET, and GENE?

4.

I'm not sure I understand the discussion around the name of the file.   Are we storing this file on the server?  I assumed the database would store the gene sets.

Reply:
1.) The comment will be stored on the gene-gene set links. I think this will mainly be used in import but we can also pull it for export. For the genesests we create we may want to include why those genes were included (typically because it was in the top X number of differentially expressed genes)

2.) I think doing CELLSET_CATEGORY.LABEL and EVERYTHING else should cover that scenario. I'm not sure about unions, I think for an mvp we'll probably stick with just labeled sets of cells but we can consider that in the future.

3.) Yes we will ignore user generated pvalues/diff exp

4.) The discussion around file name is mainly regarding export

Reply:
Are there any other features on our roadmap that would generate gene sets besides diffexp?  I thinking that our format is a bit limiting because it's csv and so we can't evolve it for future features.   At some point we might want to include more analysis features into cellxgene.  

Also, we don't have a good way to version the format.  So, if we did decide to change it in the future we wouldn't have a clear path towards backwards and forwards compatibility.

Reply:
Also regarding the format.  I can see a use case where hierarchies of gene sets would be useful.  For example, in the single cell all hands last week, we saw how NS forest was used to make a tree of genesets.   If we wanted to support that representation then we would need a way for genesets to include other genesets.   Have we decided to only support flat genesets in cellxgene?

Reply:
Regarding file names.   Will the user be allowed to choose the name when exporting genesets?

Reply:
`Are there any other features on our roadmap that would generate gene sets besides diffexp?` -- Not sure, maybe @signechambers1 knows more. 
I think this is fine for an mvp, if we decide to stop supporting this format in the future we can communicate that to users. We did consider using a json format but the users feel more comfortable with a csv.

I think for the mvp we can exclusively support flat genesets but its worth considering nested in the future. I think the db schema will be able to support that with minimal changes. 

Re file names, no. But they are welcome to rename the file when its on their machine.